### PR TITLE
Adds fees to objective computation.

### PR DESCRIPTION
**Pull request by marcovc**
Objective computation in the driver was not taking into account the order fees. This PR fixes that.

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]

File: solver/src/settlement/settlement_encoder.rs [hunk 1]
```diff
@@ -208,6 +208,20 @@ impl SettlementEncoder {
         })
     }
 
+    pub fn total_fees(&self) -> Option<U256> {
+        self.trades.iter().fold(Some(0.into()), |acc, trade| {
+            let fees = trade.order.order_creation.fee_amount;
+            Some(acc? + fees)
```
Comment (fleupold) [inline solver/src/settlement/settlement_encoder.rs:214]: does this not have to be a checked add?

File: solver/src/settlement.rs [hunk 2]
```diff
@@ -140,12 +140,34 @@ impl Settlement {
...
+    }
+
+    // For now this computes the fees of all EOA trades.
+    pub fn total_unsubsidized_fees(&self, fee_discount_factor: f64) -> U256 {
+        match self.encoder.total_unsubsidized_fees(fee_discount_factor) {
+            Some(value) => value,
+            None => {
+                tracing::error!("Overflow computing fees value for: {:?}", self);
+                0.into()
+            }
```
Comment (fleupold) [inline solver/src/settlement.rs:167]: nit: shouldn't the error handling logic be in the encoder? Or do you expect people calling encoder.total_unsubsidized_fees that don't care about the extra context?

File: solver/src/settlement.rs [hunk 3]
```diff
@@ -140,12 +140,34 @@ impl Settlement {
...
             None => {
-                tracing::error!("Overflow computing objective value for: {:?}", self);
+                tracing::error!("Overflow computing surplus value for: {:?}", self);
                 num::zero()
             }
         }
     }
 
+    // For now this computes the fees of all EOA trades.
+    pub fn total_fees(&self) -> U256 {
```
Comment (fleupold) [inline solver/src/settlement.rs:150]: nit: doesn't seem used?

File: solver/src/settlement/settlement_encoder.rs [hunk 4]
```diff
@@ -208,6 +208,20 @@ impl SettlementEncoder {
...
+        })
+    }
+
+    pub fn total_unsubsidized_fees(&self, fee_discount_factor: f64) -> Option<U256> {
```
Comment (fleupold) [inline solver/src/settlement/settlement_encoder.rs:218]: If you wanted to save some boilerplate you could get rid of this method and just divide total_fees by the fee_discount_factor in the layer above (but not a huge difference).

Comment (fleupold) [reply solver/src/settlement/settlement_encoder.rs:218]: Maybe a small unit test would be nice?

**Review by fleupold — APPROVED** [review]

File: solver/src/settlement/settlement_encoder.rs (hunk 1 above)
Comment (nlordell) [inline solver/src/settlement/settlement_encoder.rs:214]:
```suggestion
            acc?.checked_add(fees)
```

**Review by nlordell — APPROVED** [review]
Looks good. Just missing overflow checking in one spot AFAICT.

File: solver/src/settlement.rs (hunk 2 above)
Comment (marcovc) [reply solver/src/settlement.rs:167]: I was following along the existing logic around `total_surplus()` which defaults to zero for None. Do you prefer to keep it consistent or change this one to return an Option? If the latter, should we change total_surplus in a subsequent PR too?

**Comment by marcovc** [comment]
I just realized there's something wrong here. In a previous PR I assumed that fees of limit orders are always given in the sell token, can't recall how I arrived to that conclusion, probably asked someone about it. But here I'm simply adding the amounts, instead of amounts*p(sell_volume).

**Review by bh2smith — APPROVED** [review]

**Comment by marcovc** [comment]
Closed in favour of #692

**Pull request closed without merging**